gnocchi/indexer/sqlalchemy.py

```diff
@@ -568,25 +568,33 @@ def get_archive_policy(self, name):
         with self.facade.independent_reader() as session:
             return session.query(ArchivePolicy).get(name)
 
-    def update_archive_policy(self, name, ap_items):
-        with self.facade.independent_writer() as session:
-            ap = session.query(ArchivePolicy).get(name)
-            if not ap:
-                raise indexer.NoSuchArchivePolicy(name)
-            current = sorted(ap.definition,
-                             key=operator.attrgetter('granularity'))
-            new = sorted(ap_items, key=operator.attrgetter('granularity'))
-            if len(current) != len(new):
+    def update_archive_policy(self, name, ap_items, new_name=None):
+        ap = self.get_archive_policy(name)
+        if not ap:
+            raise indexer.NoSuchArchivePolicy(name)
+        current = sorted(ap.definition,
+                         key=operator.attrgetter('granularity'))
+        new = sorted(ap_items, key=operator.attrgetter('granularity'))
+        if len(current) != len(new):
+            raise indexer.UnsupportedArchivePolicyChange(
+                name, 'Cannot add or drop granularities')
+        for c, n in zip(current, new):
+            if c.granularity != n.granularity:
                 raise indexer.UnsupportedArchivePolicyChange(
-                    name, 'Cannot add or drop granularities')
-            for c, n in zip(current, new):
-                if c.granularity != n.granularity:
-                    raise indexer.UnsupportedArchivePolicyChange(
-                        name, '%s granularity interval was changed'
-                        % utils.timespan_total_seconds(c.granularity))
-            # NOTE(gordc): ORM doesn't update JSON column unless new
-            ap.definition = ap_items
-            return ap
+                    name, '%s granularity interval was changed'
+                    % utils.timespan_total_seconds(c.granularity))
+        # NOTE(gordc): ORM doesn't update JSON column unless new
+        ap.definition = ap_items
+        if new_name:
+            ap.name = new_name
+        try:
+            with self.facade.independent_writer() as session:
+                session.add(ap)
```

Comment:
Hm I'd expect this to do an insert rather than an update?

Reply:
Yeah that's weird, the object comes from a different session and the primary key might be changed so I'd except an insert too but according to the tests it seems to be doing an update in any case... I haven't digged deeper but I think I'll just revert to the old structure of the ``update_archive_policy`` method to avoid this weird behavior and the creation of 2 sessions.

Reply:
Ok, cool!